// ngx-turnstile/esm2022/lib/ngx-turnstile.component.mjs
import { Component, Input, Output, EventEmitter, Inject, afterNextRender, } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
const SCRIPT_ID = 'ngx-turnstile';
const CALLBACK_NAME = 'onloadTurnstileCallback';
export class NgxTurnstileComponent {
    elementRef;
    zone;
    document;
    siteKey;
    action;
    cData;
    theme = 'auto';
    language = 'auto';
    version = '0';
    tabIndex;
    appearance = 'always';
    retry = 'auto';
    size = 'normal';
    resolved = new EventEmitter();
    errored = new EventEmitter();
    widgetId;
    constructor(elementRef, zone, document) {
        this.elementRef = elementRef;
        this.zone = zone;
        this.document = document;
        afterNextRender(() => this.createWidget());
    }
    _getCloudflareTurnstileUrl() {
        if (this.version === '0') {
            return 'https://challenges.cloudflare.com/turnstile/v0/api.js';
        }
        throw 'Version not defined in ngx-turnstile component.';
    }
    createWidget() {
        let turnstileOptions = {
            sitekey: this.siteKey,
            theme: this.theme,
            language: this.language,
            tabindex: this.tabIndex,
            action: this.action,
            cData: this.cData,
            appearance: this.appearance,
            retry: this.retry,
            size: this.size,
            callback: (token) => {
                this.zone.run(() => this.resolved.emit(token));
            },
            'error-callback': (errorCode) => {
                this.zone.run(() => this.errored.emit(errorCode));
                // Returning false causes Turnstile to log error code as a console warning.
                return false;
            },
            'expired-callback': () => {
                this.zone.run(() => this.reset());
            },
        };
        window[CALLBACK_NAME] = () => {
            if (!this.elementRef?.nativeElement) {
                return;
            }
            this.widgetId = window.turnstile.render(this.elementRef.nativeElement, turnstileOptions);
        };
        if (this.scriptLoaded()) {
            window[CALLBACK_NAME]();
            return;
        }
        const script = this.document.createElement('script');
        script.src = `${this._getCloudflareTurnstileUrl()}?render=explicit&onload=${CALLBACK_NAME}`;
        script.id = SCRIPT_ID;
        script.async = true;
        script.defer = true;
        this.document.head.appendChild(script);
    }
    reset() {
        if (this.widgetId) {
            this.resolved.emit(null);
            window.turnstile.reset(this.widgetId);
        }
    }
    ngOnDestroy() {
        if (this.widgetId) {
            window.turnstile.remove(this.widgetId);
        }
    }
    scriptLoaded() {
        return !!this.document.getElementById(SCRIPT_ID);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: NgxTurnstileComponent, deps: [{ token: i0.ElementRef }, { token: i0.NgZone }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.0.8", type: NgxTurnstileComponent, selector: "ngx-turnstile", inputs: { siteKey: "siteKey", action: "action", cData: "cData", theme: "theme", language: "language", version: "version", tabIndex: "tabIndex", appearance: "appearance", retry: "retry", size: "size" }, outputs: { resolved: "resolved", errored: "errored" }, exportAs: ["ngx-turnstile"], ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: NgxTurnstileComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngx-turnstile',
                    template: ``,
                    exportAs: 'ngx-turnstile',
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.NgZone }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }], propDecorators: { siteKey: [{
                type: Input
            }], action: [{
                type: Input
            }], cData: [{
                type: Input
            }], theme: [{
                type: Input
            }], language: [{
                type: Input
            }], version: [{
                type: Input
            }], tabIndex: [{
                type: Input
            }], appearance: [{
                type: Input
            }], retry: [{
                type: Input
            }], size: [{
                type: Input
            }], resolved: [{
                type: Output
            }], errored: [{
                type: Output
            }] } });
